bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java

```diff
@@ -379,154 +348,216 @@ private void setRelayEvents() throws IOException {
+            request = httpClient.newRequest(url).method(method.toString()).timeout(SHELLY_API_TIMEOUT_MS,
+                    TimeUnit.MILLISECONDS);
 
-        httpResponse = HttpUtil.executeUrl(HttpMethod.GET, url, headers, null, "", SHELLY_API_TIMEOUT_MS);
-        Validate.notNull(httpResponse, "httpResponse must not be null");
-        // all api responses are returning the result in Json format. If we are getting
-        // something else it must
-        // be an error message, e.g. http result code
-        if (httpResponse.contains(APIERR_HTTP_401_UNAUTHORIZED)) {
-            throw new IOException(
-                    APIERR_HTTP_401_UNAUTHORIZED + ", set/correct userid and password in the thing/binding config");
-        }
-        if (!httpResponse.startsWith("{") && !httpResponse.startsWith("[")) {
-            throw new IOException("Unexpected http response: " + httpResponse);
+            if (!config.userId.isEmpty()) {
+                String value = config.userId + ":" + config.password;
+                request.header(HTTP_HEADER_AUTH,
+                        HTTP_AUTH_TYPE_BASIC + " " + Base64.getEncoder().encodeToString(value.getBytes()));
+            }
+            request.header(HttpHeader.ACCEPT, CONTENT_TYPE_JSON);
+            logger.trace("{}: HTTP GET for {}", thingName, url);
+
+            // Do request and get response
+            ContentResponse contentResponse = request.send();
+            apiResult = new ShellyApiResult(contentResponse);
+            String response = contentResponse.getContentAsString().replaceAll("\t", "").replaceAll("\r\n", "").trim();
+
+            // validate response, API errors are reported as Json
+            logger.trace("HTTP Response: {}", response);
+            if (contentResponse.getStatus() != HttpStatus.OK_200) {
+                throw new ShellyApiException(apiResult);
+            }
+            if (response == null || response.isEmpty() || !response.startsWith("{") && !response.startsWith("[")) {
+                throw new ShellyApiException("Unexpected response: " + response);
+            }
+        } catch (ExecutionException | InterruptedException | TimeoutException | IllegalArgumentException e) {
+            throw new ShellyApiException(apiResult, e);
         }
+        return apiResult;
+    }
 
-        logger.trace("HTTP response from {}: {}", thingName, httpResponse);
-        return httpResponse;
+    public String getControlUrlPrefix(Integer id) {
+        String uri = "";
+        if (profile.isLight || profile.isDimmer) {
+            if (profile.isDuo || profile.isDimmer) {
+                // Duo + Dimmer
+                uri = SHELLY_URL_CONTROL_LIGHT;
+            } else {
+                // Bulb + RGBW2
+                uri = "/" + (profile.inColor ? SHELLY_MODE_COLOR : SHELLY_MODE_WHITE);
+            }
+        } else {
+            // Roller, Relay
+            uri = SHELLY_URL_CONTROL_RELEAY;
+        }
+        uri = uri + "/" + id.toString();
```

Comment:
toString is called automatically
```suggestion
        uri = uri + "/" + id;
```

Reply:
changed